ocaml/xapi/stream_vdi.ml

```diff
@@ -100,6 +100,65 @@ let write_block ~__context filename buffer ofd len =
     then raise (Api_errors.Server_error (Api_errors.client_error, [ExnHelper.string_of_exn e]))
     else raise e
 
+type descriptor = {
+  length : int64;
+  status_flags : int32;
+}
+
+let flag_hole = 1l
+let flag_zero = 2l
+
+let cycle_descriptors descriptor_list offset =
+  let rec range acc a b =
+    if a > b then acc
+    else range (a::acc) (Int64.add a 1L) b
+  in
+
+  let is_empty e =
+    let has_flag flag =
+      Int32.logand e.status_flags flag = flag
+    in
+    (* We assume the destination is prezeroed, so we do not have to copy zeroed extents *)
+    (has_flag flag_hole) || (has_flag flag_zero)
+  in
+
+  let find_blocks descriptor offset increment =
+    match (is_empty descriptor) with
+      | false -> begin
+        let start_chunk = Int64.div offset increment in
+        (* If the chunk ends at the boundary don't include the next chunk *)
+        let end_chunk = Int64.div (Int64.add (Int64.add descriptor.length  offset) Int64.minus_one ) increment in
+        let chunks = range [] start_chunk end_chunk in
+        chunks, descriptor.length
+        end
+      | true -> [], descriptor.length
+  in
+  (* Only compare to most recent chunk added*)
+  let check_exists a b =
+    match b with
+    | [] -> a::b
+    | _ -> begin
+      let b_rev = List.rev b in
+      if Int64.equal a (List.hd b_rev) then b
+      else List.rev (a::b_rev)
+      end
+  in
+
+  let rec add_unique_chunks acc = function
+    | [] -> acc
```

Comment:
I think it's more efficient to not do any `List.rev`s in `check_exists` but only do one here where we return `acc`

Reply:
ok will see how I can rework it. Had been trying to always keep stuff in order (i.e. [1;2;3...]) for simplicity but I think it results in using more revs than is strictly necessary.